### PR TITLE
Introduce new 'generic.contemporary_exists' error and supporting methods

### PR DESCRIPTION
**Pull request by pond**
Includes one or two minor comment or string typo drive-by fixes and updates any instance of a primary key being generated for a test with `:id => :string`, if it were previously using `:id => false`, in passing, per http://loyaltynz.github.io/hoodoo/guides_0300_active_record.html#heading_number_3.

### CODE REVIEW

File: lib/hoodoo/errors/error_descriptions.rb [hunk 1]
```diff
@@ -97,6 +97,7 @@ def initialize( domain = nil, &block )
 
       errors_for 'generic' do
         error 'not_found',              status: 404, message: 'Resource not found',            reference: [ :ident ]
+        error 'contemporary_exists',    status: 404, message: 'Contepmorary record exists',    reference: [ :ident ]
```
Comment (bengreville) [inline lib/hoodoo/errors/error_descriptions.rb]: Typo in the message string

File: Gemfile.lock [hunk 2]
```diff
@@ -1,12 +1,11 @@
 GIT
   remote: https://github.com/pond/sdoc.git
-  revision: 5c5a85694027e946adeadfaad43cca11320f9b4d
+  revision: 35c0ad2a3920f665238709f6d92e4dc43f31aeee
```
Comment (pond) [inline Gemfile.lock]: (Drive-by maintenance `bundle update` herein)

File: README.md [hunk 3]
```diff
@@ -10,7 +10,7 @@ See [Hoodoo Guides](https://loyaltynz.github.io/hoodoo/) for extensive documenta
 
 Add the gem to your `Gemfile`:
 
-    gem 'hoodoo', '~> 1.0'
+    gem 'hoodoo', '~> 2.0'
```
Comment (pond) [inline README.md]: (Drive-by fix of oversight in V1->V2)

File: README.md [hunk 4]
```diff
@@ -65,6 +65,7 @@ Some additional higher level hand written documentation may also be present as M
 
 * [Andrew Amesbury](https://github.com/aames)
 * [Andrew Pett](https://github.com/aspett)
+* [Ben Greville](https://github.com/bengreville)
```
Comment (pond) [inline README.md]: (Drive-by: Update Contributors)

File: hoodoo.gemspec [hunk 5]
```diff
@@ -26,7 +26,7 @@ Gem::Specification.new do | s |
   s.add_development_dependency 'simplecov-rcov',   '~> 0.2'
 
   # TODO: v5 (SDoc dependency issue)
-  s.add_development_dependency 'rdoc',             '~> 4.3' # See also 'sdoc' in Gemfile
+  s.add_development_dependency 'rdoc',             '~> 5.1' # See also 'sdoc' in Gemfile
```
Comment (pond) [inline hoodoo.gemspec]: (Drive-by: Part of maintenance `bundle update`)

File: lib/hoodoo.rb [hunk 6]
```diff
@@ -12,6 +12,8 @@
 module Hoodoo
 end
 
+require 'hoodoo/monkey/monkey'
+
```
Comment (pond) [inline lib/hoodoo.rb]: In order to address the monkey patching system within the ActiveRecord extensions, the patcher code needs to be included before pretty much anything else. The actual _patches_ still only get brought in a while later.

File: lib/hoodoo/active/active_record/dated.rb [hunk 7]
```diff
@@ -289,7 +300,7 @@ def dated_at( date_time = Time.now, unquoted_column_names: nil )
         #
         # +unquoted_column_names+:: (Optional) An Array of Strings giving one
         #                           or more column names to use for the query.
-        #                           If omitted, all model attribtues are used
+        #                           If omitted, all model attributes are used
```
Comment (pond) [inline lib/hoodoo/active/active_record/dated.rb]: (Drive-by: Fix typo)

File: lib/hoodoo/monkey.rb [hunk 8]
```diff
@@ -10,6 +10,9 @@
 ########################################################################
 
 require 'hoodoo/monkey/monkey'
+
+require 'hoodoo/monkey/patch/active_record_dated_finder_additions'
+require 'hoodoo/monkey/patch/active_record_manually_dated_finder_additions'
+require 'hoodoo/monkey/patch/datadog_traced_amqp'
 require 'hoodoo/monkey/patch/newrelic_middleware_analytics'
 require 'hoodoo/monkey/patch/newrelic_traced_amqp'
-require 'hoodoo/monkey/patch/datadog_traced_amqp'
```
Comment (pond) [inline lib/hoodoo/monkey.rb]: (Drive-by: Alphabetic ordering for simplicity, instead of arbitrary order implying some ordering significance)

File: lib/hoodoo/monkey/patch/datadog_traced_amqp.rb [hunk 9]
```diff
@@ -8,7 +8,7 @@
 #
 #           See Hoodoo::Monkey::Patch::DatadogTracedAMQP for more.
 # ----------------------------------------------------------------------
-#           22-June-2017 (JRW): Created.
+#           22-Jun-2017 (JRW): Created.
```
Comment (pond) [inline lib/hoodoo/monkey/patch/datadog_traced_amqp.rb]: (Drive-by: Date format fix)

File: docs/api_specification/README.md [hunk 10]
```diff
@@ -1346,3 +1349,4 @@ It is likely to be helpful if you augment this with your own selection of search
 | 2016-12-06 | Release 4 | ADH    | Describe new framework-level search/query strings of `created_after` and `created_before`. Mention potential for using this instead of very large offset values. |
 | 2017-08-17 | Release 5 | ADH    | Describe new standard optional resource field `created_by`, for resource fingerprints. |
 | 2017-10-13 | Release 6 | ADH    | Hoodoo 2; describe new framework-level search/query string of `created_by`. |
+| 2017-11-02 | Release 7 | ADH    | Add +generic.contepmorary_exists+ error and associated information. |
```
Comment (bengreville) [inline docs/api_specification/README.md]: Minor, but same typo exists here likely from copy paste.

File: lib/hoodoo/active/active_record/finder.rb [hunk 11]
```diff
@@ -246,61 +270,138 @@ def acquire( ident )
...
+        # Example, following on from those for #acquire_in:
+        #
+        #     def show( context )
+        #       resource = SomeModel.acquire_in_and_update( context )
+        #       return if context.response.halt_processing? # Or just use 'if resource.nil?'
+        #
+        #       # ...else render...
+        #     end
+        #
+        def acquire_in_and_update( context )
```
Comment (mbdietrich) [inline lib/hoodoo/active/active_record/finder.rb]: The general approach seems good, though I don't agree with the `acquire_in_and_update` name (sounds like it'd find an object, then update it). Not a major complaint though.

`acquire_in_with_errors` or `acquire_in!` maybe?

File: lib/hoodoo/errors/error_descriptions.rb [hunk 12]
```diff
@@ -97,6 +97,7 @@ def initialize( domain = nil, &block )
 
       errors_for 'generic' do
         error 'not_found',              status: 404, message: 'Resource not found',            reference: [ :ident ]
+        error 'contemporary_exists',    status: 404, message: 'Contemporary record exists',    reference: [ :ident ]
```
Comment (mbdietrich) [inline lib/hoodoo/errors/error_descriptions.rb]: More accurate maybe, "Record exists, but not at the given datetime" or something like that?

File: docs/api_specification/README.md (hunk 10 above)
Comment (pond) [reply docs/api_specification/README.md]: Argh! Thanks, well spotted.

File: lib/hoodoo/active/active_record/finder.rb (hunk 11 above)
Comment (pond) [reply lib/hoodoo/active/active_record/finder.rb]: `acquire_in!` is an excellent suggestion, will update.

Comment (pond) [reply lib/hoodoo/active/active_record/finder.rb]: (Implies future addition of create/update/delete "!" variants with the same intent, for both normal and manual dating. There's a lot of error handling and not-found boilerplate presently).

Comment (pond) [reply lib/hoodoo/active/active_record/finder.rb]: ...@mbdietrich https://github.com/LoyaltyNZ/hoodoo/issues/225 captures this.

File: lib/hoodoo/errors/error_descriptions.rb (hunk 12 above)
Comment (pond) [reply lib/hoodoo/errors/error_descriptions.rb]: I can't currently think of something appropriately terse that's better than the currently written message, which I _think_ is OK given:

* It's the code that's machine-readable, the message is just for humans
* Our intended use for this is always in concert with 'not found', so you'll have two entries - "Resource not found; Contemporary record exists"
* If a service author had a good reason to _only_ use `generic.contemporary_exists` they can make a judgement call on whether their use context means the default message is appropriate, or if they want to override with something else.
